### PR TITLE
Added synergy formula to build from sources at latest version (1.6.3)

### PR DESCRIPTION
**Pull request by guyzmo**
- installs .app and CLI tools
 - included patch to fix build issue
 - included patch to not codesign the app

Signed-off-by: Guyzmo <guyzmo+github@m0g.net>

### CODE REVIEW

File: Library/Formula/synergy.rb [hunk 1]
```diff
@@ -0,0 +1,75 @@
+class Synergy < Formula
+  homepage 'http://synergy-project.org'
```
Comment (DomT4) [inline Library/Formula/synergy.rb]: Double quotes everywhere.

File: Library/Formula/synergy.rb [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+  url 'https://github.com/synergy/synergy/archive/1.6.3.tar.gz'
+  sha1 'c4c26ee983f0684a5cbe45520adc0b50a71a3524'
```
Comment (DomT4) [inline Library/Formula/synergy.rb]: Should be ` sha256 `.

File: Library/Formula/synergy.rb [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+  depends_on "cmake" => :build
+  depends_on "qt5" => :build
```
Comment (DomT4) [inline Library/Formula/synergy.rb]: No runtime usage of ` qt5 `?

File: Library/Formula/synergy.rb [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+
+  def install
+    ENV.deparallelize  # if your formula fails when building in parallel
```
Comment (DomT4) [inline Library/Formula/synergy.rb]: Cut the code comment.

File: Library/Formula/synergy.rb [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+
+    system './hm.sh', 'conf', '-g1', '--mac-sdk', '10.10', '--mac-identity', 'test'
```
Comment (DomT4) [inline Library/Formula/synergy.rb]: Can't presume the SDK available is going to be ` 10.10 `.

File: Library/Formula/synergy.rb [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+
+    bin.install 'bin/synergyc'
+    bin.install 'bin/synergyd'
+    bin.install 'bin/synergys'
+    bin.install 'bin/syntool'
+
+    prefix.install 'bin/Synergy.app'
+  end
+
+  patch :DATA
```
Comment (DomT4) [inline Library/Formula/synergy.rb]: This should be above ` def install `, not below it. Please also explain why the patch is necessary, briefly, in formula, and link to an upstream commit/bug report/PR submission.

File: Library/Formula/synergy.rb [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+
+  test do
+    system "#{bin}/synergys", '--version'
+  end
+
+  def caveats; <<-EOS
```
Comment (DomT4) [inline Library/Formula/synergy.rb]: Can kill this. It'll be handled by ` brew ` automatically.

File: Library/Formula/synergy.rb [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+    bin.install 'bin/synergys'
+    bin.install 'bin/syntool'
+
+    prefix.install 'bin/Synergy.app'
+  end
+
+  patch :DATA
+
+  test do
+    system "#{bin}/synergys", '--version'
```
Comment (DomT4) [inline Library/Formula/synergy.rb]: Can this test be improved at all? See cmake.rb for an example of an application formula with a good test and tinyxml2.rb for an example of a library formula with a good test. Thanks!

File: Library/Formula/synergy.rb (hunk 1 above)
Comment (guyzmo) [inline Library/Formula/synergy.rb]: !@#$! should make a single rule for both cask and brew :-s

File: Library/Formula/synergy.rb (hunk 3 above)
Comment (guyzmo) [inline Library/Formula/synergy.rb]: nope, qt5 is copied within the resources of the .app.

File: Library/Formula/synergy.rb (hunk 5 above)
Comment (guyzmo) [inline Library/Formula/synergy.rb]: could you hint me how can I get the current SDK value, please? ;-)

Comment (DomT4) [inline Library/Formula/synergy.rb]: ` macosx#{MacOS.version} `

**Comment by mikemcquaid** [comment]
What's the advantage in building this from source compared to the Homebrew Cask version?

File: Library/Formula/synergy.rb (hunk 1 above)
Comment (DomT4) [inline Library/Formula/synergy.rb]: We don't control the Cask's style choices, for better or worse. They're a separate project that just hop onto Homebrew's tap mechanism, essentially. Homebrew uses double quotes everywhere, attempting to follow the Ruby Style Guide.

Comment (guyzmo) [inline Library/Formula/synergy.rb]: ok, then I'll complain next time I write a cask :-)

File: Library/Formula/synergy.rb (hunk 2 above)
Comment (guyzmo) [inline Library/Formula/synergy.rb]: :+1:

sorry for the error: I have recycled an older version of the formula, fixing that.

File: Library/Formula/synergy.rb (hunk 4 above)
Comment (guyzmo) [inline Library/Formula/synergy.rb]: :+1:

sorry for the error: I have recycled an older version of the formula

File: Library/Formula/synergy.rb (hunk 7 above)
Comment (guyzmo) [inline Library/Formula/synergy.rb]: :+1:

File: Library/Formula/synergy.rb (hunk 6 above)
Comment (guyzmo) [inline Library/Formula/synergy.rb]: :+1:

File: Library/Formula/synergy.rb (hunk 8 above)
Comment (guyzmo) [inline Library/Formula/synergy.rb]: well… we might make a quite complicated test by launching the server, then the client, then testing they connect by parsing both outputs :-s But that's quite a complicated thing to do…

I could expand the tests to check that all binaries have been correctly compiled, with --version on all.

Comment (guyzmo) [inline Library/Formula/synergy.rb]: hum on the following test:

http://bot.brew.sh/job/Homebrew%20Pull%20Requests/23520/version=yosemite/testReport/brew-test-bot/yosemite/test_synergy/

I don't see how can I make a good test on a server that has no UI, for an app that needs an UI to work :-/

**Comment by guyzmo** [comment]
@mikemcquaid I actually did not know about the synergykm builds, and anyway it's not within the main caskroom, but as a tap. I tend to believe it's worth having it within brew, as it's a FLOSS project, which offers CLI tools as well as GUI tools.

**Comment by guyzmo** [comment]
N.B.: I'm actually paying/supporting synergy but still prefer to get it through homebrew ;-)

File: Library/Formula/synergy.rb (hunk 8 above)
Comment (guyzmo) [inline Library/Formula/synergy.rb]: I removed the test as I don't see how I'll be able to make it work through travis which does not feature an UI.

File: Library/Formula/synergy.rb (hunk 5 above)
Comment (guyzmo) [inline Library/Formula/synergy.rb]: from Mr Travis:
> * Don't need to interpolate "MacOS.version" with system

File: Library/Formula/synergy.rb [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+    bin.install "bin/synergyc"
+    bin.install "bin/synergyd"
+    bin.install "bin/synergys"
+    bin.install "bin/syntool"
+
+    prefix.install "bin/Synergy.app"
+  end
+end
+
+# First patch is a fix for building for OSX. There is an error where the default
```
Comment (DomT4) [inline Library/Formula/synergy.rb]: Should go above the ` patch :DATA ` line :)

File: Library/Formula/synergy.rb (hunk 5 above)
Comment (DomT4) [inline Library/Formula/synergy.rb]: Heh. I forgot it'd get punchy about that. I presume it wants you to change it to:

```ruby
system "./hm.sh", "conf", "-g1", "--mac-sdk", MacOS.version, "--mac-identity", "test"
```

File: Library/Formula/synergy.rb [hunk 10]
```diff
@@ -0,0 +1,63 @@
+class Synergy < Formula
+  homepage "http://synergy-project.org"
+  url "https://github.com/synergy/synergy/archive/1.6.3.tar.gz"
+  sha256 "93b1965b8e0cfc55cc654aab7053a58b9e730d09e68bbc520be32353a01dce1a"
+
+  depends_on "cmake" => :build
+  depends_on "qt5" => :build
+
+  patch :DATA
```
Comment (mikemcquaid) [inline Library/Formula/synergy.rb]: Can you please submit this patch to the upstream developers of this project and add a link to the upstream patch submission and explanation of why the patch is needed in a comment in the formula file. Thanks!

File: Library/Formula/synergy.rb (hunk 5 above)
Comment (guyzmo) [inline Library/Formula/synergy.rb]: :+1: changing that, thank you ;-)

File: Library/Formula/synergy.rb (hunk 10 above)
Comment (guyzmo) [inline Library/Formula/synergy.rb]: A patch is in process on the upstream: https://github.com/synergy/synergy/pull/4146

I'll push another patch for not signing the .app as well. Then we might end up with no patches at all for 1.6.4 ;-)

Comment (mikemcquaid) [inline Library/Formula/synergy.rb]: We'll need all patches to be merged before including this.

Comment (guyzmo) [inline Library/Formula/synergy.rb]: Well, the patches are still being reviewed, and won't be included before release 1.6.4.

That means we could keep the 1.6.3 formula until it's deprecated by the new version. 

I'm getting notifications of new releases from synergy, so I'll be able to update the formula as soon as it is released.

Comment (mikemcquaid) [inline Library/Formula/synergy.rb]: I'd probably rather we waited until `1.6.4` in that case, thanks.

Comment (guyzmo) [inline Library/Formula/synergy.rb]: err… ok?

why not integrate a formula that is ready and works, with only a couple of minor patches?

Comment (mikemcquaid) [inline Library/Formula/synergy.rb]: We prefer to wait for patches to get merged before adding them to formulae, particularly new formulae. I don't personally consider it ready and working as long as it requires patches that upstream hasn't accepted. Once the patches are accept we can consider using the upstream commits as the patches.

**Comment by mikemcquaid** [comment]
`audit` failing: http://bot.brew.sh/job/Homebrew%20Pull%20Requests/23531/version=yosemite/testReport/junit/brew-test-bot/yosemite/audit_synergy___strict/

**Comment by guyzmo** [comment]
cf the [previous build](http://bot.brew.sh/job/Homebrew%20Pull%20Requests/23520/version=yosemite/testReport/brew-test-bot/yosemite/test_synergy/), the test I wrote failed because it's running on a system lacking a WindowServer, and I'm not sure how I can make a good test for that.

File: Library/Formula/synergy.rb (hunk 10 above)
Comment (guyzmo) [inline Library/Formula/synergy.rb]: hum ok! 

is there a good tap I could push it to? 
or would it be worth having me push it to my own tap, until it's patched?

File: Library/Formula/synergy.rb (hunk 9 above)
Comment (guyzmo) [inline Library/Formula/synergy.rb]: :+1: ! 
gonna fix and squash that in

File: Library/Formula/synergy.rb (hunk 10 above)
Comment (mikemcquaid) [inline Library/Formula/synergy.rb]: Yeh, that's cool with us.

**Comment by jrolfs** [comment]
@guyzmo did you end up pushing a tap somewhere?

**Comment by DomT4** [comment]
There's a new release available here, which would resolve Mike's concern, if you're still interested in this formula.

**Comment by jrolfs** [comment]
Yeah they're up to 1.7.4 I believe but I checked yesterday and it still [needs the 10.10 patch](https://github.com/synergy/synergy/pull/4146) as far as I can tell. Would you consider allowing the patched version? If so I can help push this through.

**Comment by DomT4** [comment]
We tend to require upstream to accept the fix prior to merge. It tends to be the case that otherwise we end up supporting a patch for months or years whilst we wait on upstream. Is there any chance we can enquire upstream about the status of the PR?

**Comment by jrolfs** [comment]
It looks like it may just be some conflicts that need to be worked out, hopefully someone with write access can take care of them soon.

**Comment by DomT4** [comment]
Upstream hasn't touched the bug report in four months despite the comment, so I'm going to close here for now :cry:. If we can get the issue fixed upstream we're more than happy to review a fresh PR for this formula.

Thank for for the submission @guyzmo, and thank you @jrolfs for trying to get an upstream response on this.

**Pull request closed without merging**